lingvo/tasks/asr/ctc_model.py

```diff
@@ -34,212 +38,234 @@ class CTCModel(base_model.BaseTask):
+      input_batch_src = self.frontend.FProp(theta.frontend, input_batch_src)
+
+    if p.input_stacking_tpl:
+      inputs = tf.squeeze(input_batch_src.src_inputs, [-1])
+      rnn_padding = tf.expand_dims(input_batch_src.paddings, -1)
+      inputs, rnn_padding = self.input_stacking.FProp(inputs, rnn_padding)
       input_batch_src = py_utils.NestedMap(
-          src_inputs=audio, paddings=tf.zeros_like(audio))
-      input_batch_src = frontend.FPropDefaultTheta(input_batch_src)
+        src_inputs=tf.expand_dims(inputs, -1),
+        paddings=tf.squeeze(rnn_padding, [-1])
+      )
 
-      encoder_outputs = self.FPropDefaultTheta() # _FProp()
-      decoder_outputs = self.decoder.BeamSearchDecode(encoder_outputs)
-      topk = self._GetTopK(decoder_outputs)
+    encoder_out = self.encoder.FProp(theta.encoder, input_batch_src)
 
-      feeds = {'wav': wav_bytes}
-      fetches = {
-          'hypotheses': topk.decoded,
-          'scores': topk.scores,
-          'src_frames': input_batch_src.src_inputs,
-          'encoder_frames': encoder_outputs.encoded
-      }
+    encoded, padding = self.project_to_vocab_size(encoder_out.encoded, encoder_out.padding)
+    outputs = py_utils.NestedMap(encoded=encoded, padding=padding)
+    return outputs
 
-      return fetches, feeds
+  def _DecodeCTC(self, output_batch):
+    # swap row 0 and row 73 because decoder assumes blank is at 0,
+    # however we set blank = 73
+    tok_logits = output_batch.encoded  # (T, B, F)
+    idxs = list(range(tok_logits.shape[-1]))
+    idxs[0] = self.params.blank_index
+    idxs[self.params.blank_index] = 0
+    tok_logits = tf.stack([tok_logits[:, :, idx] for idx in idxs], axis=-1)
 
-  @classmethod
-  def FPropMeta(cls, params, *args, **kwargs):
-    raise NotImplementedError("No FPropMeta available.")
+    (decoded,), neg_sum_logits = tf.nn.ctc_beam_search_decoder(
+      tok_logits,
+      py_utils.LengthsFromBitMask(output_batch.padding, 0),
+      beam_width=100
+    )
+
+    dense_dec = tf.sparse_to_dense(
+      decoded.indices, decoded.dense_shape, decoded.values, default_value=self.params.blank_index
+    )
+
+    INVALID = tf.constant(self.params.blank_index, tf.int64)
+    bitMask = tf.cast(tf.math.equal(dense_dec, INVALID), tf.float32)  # (B, T)
+
+    decoded_seq_lengths = py_utils.LengthsFromBitMask(tf.transpose(bitMask), 0)
+
+    hyp_str = self.input_generator.IdsToStrings(
+      tf.cast(dense_dec, tf.int32), decoded_seq_lengths
+    )
+
+    # Some predictions have start and stop tokens predicted, we dont want to include
+    # those in WER calculation
+    hyp_str = tf.strings.regex_replace(hyp_str, '(<unk>)+', '')
+    hyp_str = tf.strings.regex_replace(hyp_str, '(<s>)+', '')
+    hyp_str = tf.strings.regex_replace(hyp_str, '(</s>)+', '')
+    return py_utils.NestedMap(sparse_ids=decoded, transcripts=hyp_str)
+
+  def _CalculateErrorRates(self, dec_outs_dict, input_batch):
+    sparse_gt_ids = tf.cast(py_utils.SequenceToSparseTensor(
```

Comment:
Dead code?

Reply:
It's used in line 111. Why do you say it's dead?

Reply:
I mean that sparse_gt_ids is an unused variable, to clarify.